### PR TITLE
Olwa pr

### PR DESCRIPTION
**Pull request by oliver-walker-hms**

### CODE REVIEW

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java [hunk 1]
```diff
@@ -168,6 +168,12 @@ public static synchronized ArrayList getFifoNextSpanData(
     } else {
       startTimeTrackerMs = FileAccessManager.readFileToString(timeMarkerFileName);
       startTimeTrackerMsLong = Long.parseLong(startTimeTrackerMs);
+      // If the file is corrupted, holds no data
+      if (startTimeTrackerMs.equalsIgnoreCase("empty")) {
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]: wouldnt line 169 be setting startTimeTrackerMs to null in the case where the file doesnt exist? And if it does exist it will be whatever is in that file.

Comment (oliver-walker-hms) [reply src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]: I changed this to a draft because it needed to be tested, you're right the exception is thrown when a null or empty value is being parse as a long. Whenever I upload an empty file to the device that file gets deleted, so it has been difficult to test.

Comment (TomKimsey) [reply src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]: Sorry, didnt see the draft, just saw the review request.

Comment (oliver-walker-hms) [reply src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]: this is now ready for review.

**Review by alexjhawk — CHANGES_REQUESTED** [review]
Just to add context to this PR (@oliver-walker-hms and I spoke already), this exception should be handled by the calling application (to allow for logging and retry mechanisms). The calling application can then trigger a new time tracker to be created by the `startNewTimeTracker` boolean parameter

**Comment by oliver-walker-hms** [comment]
This PR is in draft again. I will post here when it is ready for review

File: src/com/hms_networks/americas/sc/historicaldata/CorruptedTimeTrackerException.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package com.hms_networks.americas.sc.historicaldata;
```
Comment (alexjhawk) [inline src/com/hms_networks/americas/sc/historicaldata/CorruptedTimeTrackerException.java:1]: I think the exceptions should be added in their own commit

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java [hunk 3]
```diff
@@ -150,26 +264,38 @@ public static synchronized ArrayList getFifoNextSpanDataAllGroups(boolean startN
    * @param includeTagGroupD if tag group D data should be included
    * @return historical log data
    * @throws IOException if unable to read or write files
+   * @throws TimeTrackerUnrecoverableException if both time tracking files are corrupted
+   * @throws CorruptedTimeTrackerException if the primary time tracking file is corrupted
    */
   public static synchronized ArrayList getFifoNextSpanData(
       boolean startNewTimeTracker,
```
Comment (alexjhawk) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java:285]: Doesn't appear that `startNewTimeTracker` is used anymore. I think it should either still be used to force new time trackers to be created, or be removed. I lean more towards still handling that boolean, because I do see a use case where a connector could force a time tracker reset. In the past, I have told customers how to shut off a connector, and manually delete the tracker files so they are recreated at current time.

**Review by alexjhawk — CHANGES_REQUESTED** [review]

File: src/com/hms_networks/americas/sc/historicaldata/CorruptedTimeTrackerException.java (hunk 2 above)
Comment (TomKimsey) [reply src/com/hms_networks/americas/sc/historicaldata/CorruptedTimeTrackerException.java:1]: im fine with them being left in as this commit is the commit that makes use of them. They would be fine in their own commits though.

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java (hunk 3 above)
Comment (oliver-walker-hms) [reply src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java:285]: startNewTimeTracker is still used on line 297. I was thinking that behavior of passing startNewTimeTracker as true rather then rely on exceptions in the case the files don't exist makes more sense. The exception handling is really just for error's.

Comment (alexjhawk) [reply src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java:285]: missed that

File: src/com/hms_networks/americas/sc/historicaldata/CorruptedTimeTrackerException.java (hunk 2 above)
Comment (alexjhawk) [reply src/com/hms_networks/americas/sc/historicaldata/CorruptedTimeTrackerException.java:1]: I'm fine either way,. Adding them in their own commit I think would just provide a slightly better/cleaner git log.

Comment (oliver-walker-hms) [reply src/com/hms_networks/americas/sc/historicaldata/CorruptedTimeTrackerException.java:1]: I went ahead and put them in a separate commit

**Review by alexjhawk — DISMISSED** [review]

**Review by TomKimsey — COMMENTED** [review]
Names and comments reference primary and backup/fallback tracker. This is very misleading as they are both the primary tracker, it just switches between the two.

**Comment by oliver-walker-hms** [comment]
@TomKimsey I didn't have a good name for these since they could either be primary or secondary depending on the point in time. Would timeMarkerFile1 and timeMarkerFile2 work as an alternative?

**Comment by TomKimsey** [comment]
> @TomKimsey I didn't have a good name for these since they could either be primary or secondary depending on the point in time. Would timeMarkerFile1 and timeMarkerFile2 work as an alternative?

yes, anything that indicates that they are both being used as the primary is fine

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java [hunk 4]
```diff
@@ -26,11 +26,27 @@
   /** Boolean flag indicating if string history data should be included in queue data. */
   private static boolean stringHistoryEnabled = false;
 
-  /** File path for time marker */
-  private static final String timeMarkerFileName =
+  /**
+   * Boolean flag indicating if the first time marker file is the primary or secondary time marker
+   * file.
+   */
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]: comment seems stale

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java [hunk 5]
```diff
@@ -26,11 +26,27 @@
...
+  private static boolean useTimeTrackerFile1Primary = true;
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]:
```suggestion
  private static boolean isFile1CurrTimeTrackerFile = true;
```

Comment (TomKimsey) [reply src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]: this might more accuratly represent this variable. primary/secondary naming makes it seem like one file is always deemed primary rather where curr or current or even active indicate there is a time element associated with the tracking file used.

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java [hunk 6]
```diff
@@ -140,6 +158,104 @@ public static synchronized ArrayList getFifoNextSpanDataAllGroups(boolean startN
...
+    } else if (file2TimeFailed) {
+      useTimeTrackerFile1Primary = true;
+      throw new CorruptedTimeTrackerException(errorMessageSingle);
+    } else {
+      // Both files are in a good state, compare for more recent time.
+      if (file1Time > file2Time) {
+        useTimeTrackerFile1Primary = true;
+      } else {
+        useTimeTrackerFile1Primary = false;
+      }
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]: using a ternary might clean this section up a bit.

**Review by TomKimsey — CHANGES_REQUESTED** [review]

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java (hunk 6 above)
Comment (TomKimsey) [reply src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]: not necessary though

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataConstants.java [hunk 7]
```diff
@@ -50,6 +50,15 @@
   /** Name of EBD string export call file */
   static final String QUEUE_EBD_STRING_FILE_NAME = "histStringDataEBD";
 
-  /** Name of historical data queue time tracker file */
-  static final String QUEUE_TIME_FILE_NAME = "histDataTime";
+  /**
+   * Name of historical data queue time tracker file 1. Can be the primary or secondary time
+   * tracking file.
+   */
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataConstants.java]:
```suggestion
  /**
   * Filename of historical data queue time tracker file 1.
   */
```

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataConstants.java [hunk 8]
```diff
@@ -50,6 +50,15 @@
...
+  static final String QUEUE_TIME_FILE_1_NAME = "histDataTime1";
+
+  /**
+   * Name of historical data queue time tracker file 2. Can be the primary or secondary time
+   * tracking file.
+   */
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataConstants.java]:
```suggestion
  /**
   * Filename of historical data queue time tracker file 2.
   */
```

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java [hunk 9]
```diff
@@ -26,11 +26,27 @@
   /** Boolean flag indicating if string history data should be included in queue data. */
   private static boolean stringHistoryEnabled = false;
 
-  /** File path for time marker */
-  private static final String timeMarkerFileName =
+  /**
+   * Boolean flag indicating if reads are from the first or second time tracking file. If set to
+   * false, reads are from the second time tracking file.
+   */
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]:
```suggestion
  /**
   * Boolean flag indicating if reads are from QUEUE_TIME_FILE_1_NAME or QUEUE_TIME_FILE_2_NAME. If set to
   * false, reads are from QUEUE_TIME_FILE_2_NAME.
   */
```

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java [hunk 10]
```diff
@@ -140,6 +158,100 @@ public static synchronized ArrayList getFifoNextSpanDataAllGroups(boolean startN
         includeTagGroupD);
   }
 
+  /**
+   * Determine which of the time tracking files is the primary file. Recover from corrupted state if
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]:
```suggestion
   * Determine which of the time tracking files is the current file. Recover from corrupted state if
```

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java [hunk 11]
```diff
@@ -140,6 +158,100 @@ public static synchronized ArrayList getFifoNextSpanDataAllGroups(boolean startN
...
+   * necessary. This function should be called before the isFile1CurrTimeTrackerFile flag is first
+   * used.
+   *
+   * @throws TimeTrackerUnrecoverableException if both time tracking files are corrupted
+   * @throws IOException if unable to read or write files
+   * @throws CorruptedTimeTrackerException if the primary time tracking file is corrupted
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]:
```suggestion
   * @throws CorruptedTimeTrackerException if the one of the tracking files is corrupted
```

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java [hunk 12]
```diff
@@ -140,6 +158,100 @@ public static synchronized ArrayList getFifoNextSpanDataAllGroups(boolean startN
...
+  /**
+   * Retrieve the start time if it exists, otherwise create a new start time and return the new
+   * start time.
+   *
+   * @param startNewTimeTracker set to true if a new time tracker file is needed.
+   * @param readFile the file path of the time tracking file to read from
+   * @param writeFile the file path of the time tracking file to write to
+   * @return the long representation of the stored time
+   * @throws IOException if unable to read or write files
+   * @throws CorruptedTimeTrackerException if the primary time tracking file is corrupted
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]:
```suggestion
   * @throws CorruptedTimeTrackerException if one of the tracking files is corrupted
```

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java [hunk 13]
```diff
@@ -150,26 +262,38 @@ public static synchronized ArrayList getFifoNextSpanDataAllGroups(boolean startN
    * @param includeTagGroupD if tag group D data should be included
    * @return historical log data
    * @throws IOException if unable to read or write files
+   * @throws TimeTrackerUnrecoverableException if both time tracking files are corrupted
+   * @throws CorruptedTimeTrackerException if the primary time tracking file is corrupted
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]:
```suggestion
   * @throws CorruptedTimeTrackerException if one of the tracking files is corrupted
```

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java [hunk 14]
```diff
@@ -150,26 +262,38 @@ public static synchronized ArrayList getFifoNextSpanDataAllGroups(boolean startN
...
-      startTimeTrackerMs = Long.toString(startTimeTrackerMsLong);
-      FileAccessManager.writeStringToFile(timeMarkerFileName, startTimeTrackerMs);
+      throws IOException, TimeTrackerUnrecoverableException, CorruptedTimeTrackerException {
+
+    if (!hasInitTime) {
+      initTimeTrackerFiles();
+    }
+
+    /*
+     * The primary and secondary time tracking files will swap on every iteration of grabbing new
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]:
```suggestion
     * The two time tracking files will swap on every iteration of grabbing new
```

File: src/com/hms_networks/americas/sc/package.html [hunk 15]
```diff
@@ -14,7 +14,7 @@
 This library is dependent on the HMS Networks MU Americas Solution Center library for file
 utilities, found at https://github.com/hms-networks/sc-flexy-logger-lib.
 
-@version 2.1
+@version 2.2
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/package.html:17]: spelling error in commit title

I assume you are planning for a release so this (https://teams.microsoft.com/l/entity/com.microsoft.teamspace.tab.wiki/tab::78561cc3-4c3b-4047-b463-167a3831b5ba?context=%7B%22subEntityId%22%3A%22%7B%5C%22pageId%5C%22%3A5%2C%5C%22sectionId%5C%22%3A19%2C%5C%22origin%5C%22%3A2%7D%22%2C%22channelId%22%3A%2219%3Accb41d2097924ef6aa065081838e1e83%40thread.skype%22%7D&tenantId=5c7c1590-4488-4e42-bc9c-15218f8ac994) should be followed

**Review by TomKimsey — CHANGES_REQUESTED** [review]

File: src/com/hms_networks/americas/sc/package.html (hunk 15 above)
Comment (oliver-walker-hms) [reply src/com/hms_networks/americas/sc/package.html:17]: Fixed

Comment (TomKimsey) [reply src/com/hms_networks/americas/sc/package.html:17]: spelling error in commit title

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java [hunk 16]
```diff
@@ -105,7 +121,7 @@ public static void setStringHistoryEnabled(boolean stringHistoryEnabled) {
    * @return true if time tracker file exists
    */
   public static boolean doesTimeTrackerExist() {
-    return new File(timeMarkerFileName).isFile();
+    return new File(timeMarkerFile1Name).isFile();
```
Comment (alexjhawk) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]: Shouldn't this method be updated to handle the new multiple file logic?

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java [hunk 17]
```diff
@@ -115,7 +131,7 @@ public static boolean doesTimeTrackerExist() {
    * @return time tracker file value
    */
   public static long getCurrentTimeTrackerValue() throws IOException {
-    String startTimeTrackerMs = FileAccessManager.readFileToString(timeMarkerFileName);
+    String startTimeTrackerMs = FileAccessManager.readFileToString(timeMarkerFile1Name);
```
Comment (alexjhawk) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataQueueManager.java]: Same comment: Shouldn't this method be updated to handle the new multiple file logic?

**Review by alexjhawk — CHANGES_REQUESTED** [review]

File: src/com/hms_networks/americas/sc/package.html (hunk 15 above)
Comment (TomKimsey) [reply src/com/hms_networks/americas/sc/package.html:17]: Still a spelling error in commit title
RELEASE: Incremented version to 2.2

File: src/com/hms_networks/americas/sc/historicaldata/HistoricalDataConstants.java [hunk 18]
```diff
@@ -50,6 +50,9 @@
   /** Name of EBD string export call file */
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/historicaldata/HistoricalDataConstants.java:50]: commit message should get updated to 

```
Adds automated recovery from corrupted file. Uses a second
time tracker file to prevent lost data due to file corruption. Uses
current time in the case both files are lost.
```

**Review by TomKimsey — CHANGES_REQUESTED** [review]

**Review by alexjhawk — APPROVED** [review]

**Review by TomKimsey — APPROVED** [review]
